### PR TITLE
Report constraint violation exceptions for Commands and Events in a unified way

### PR DESCRIPTION
**Pull request by armiol**
`InvalidCommandException` and `InvalidEventException` are used to report the invalid `Command` and `Event` messages which fields violate constraints.

Before this PR the creation of these exceptions was implemented in a very similar way, with some of the code duplicated.

This PR introduces an `@Internal` abstract helper class (`ConstraintViolations. ExceptionBuilder`) aimed to build these exceptions in a unified way.

### CODE REVIEW

**Comment by armiol** [comment]
@alexander-yevsyukov PTAL.

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java [hunk 1]
```diff
@@ -128,4 +137,89 @@ public static String toText(String format, Iterable<ConstraintViolation> violati
...
+
+        protected abstract String getErrorText();
+
+
+        protected abstract Map<String, Value> getMessageTypeAttribute(Message message);
+
+        protected abstract E createException(String exceptionMsg, M message, Error error);
+
+        private String formatExceptionMessage() {
+            return String.format("%s. Message class: %s. " +
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: Some parts of our code use statically imported `format()`. I think we need to standardize. I've never seen anybody doing this, but it makes the code a big shorter. Let's discuss.

File: server/src/main/java/org/spine3/server/command/InvalidCommandException.java [hunk 2]
```diff
@@ -46,44 +49,26 @@
 
     private static final long serialVersionUID = 0L;
 
-    private static final String MSG_VALIDATION_ERROR = "Command message does match validation constrains.";
+    private static final String MSG_VALIDATION_ERROR = "Command message does match " +
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/command/InvalidCommandException.java]: Does not?

File: server/src/main/java/org/spine3/server/command/InvalidCommandException.java [hunk 3]
```diff
@@ -117,4 +102,48 @@ public static Error unknownTenantError(Message commandMessage, String errorText)
...
+            return MSG_VALIDATION_ERROR;
+        }
+
+        @Override
+        protected Map<String, Value> getMessageTypeAttribute(Message commandMessage) {
+            return commandTypeAttribute(commandMessage);
+        }
+
+        @Override
+        protected InvalidCommandException createException(String exceptionMsg, Command message,
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/command/InvalidCommandException.java]: Consider having all parameters in one column if they don't fit into one line.

File: server/src/main/java/org/spine3/server/event/InvalidEventException.java [hunk 4]
```diff
@@ -41,7 +43,7 @@
     private static final long serialVersionUID = 0L;
 
     private static final String MSG_VALIDATION_ERROR = "Event message does match " +
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/InvalidEventException.java]: Does not?

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java [hunk 5]
```diff
@@ -128,4 +137,89 @@ public static String toText(String format, Iterable<ConstraintViolation> violati
...
+     * A helper class for building exceptions used to report invalid {@code Message}s,
+     * which have fields that violate validation constraint(s).
+     *
+     * @param <E> type of {@code Exception} to build
+     * @param <M> type of the {@code Message}
+     * @param <C> type of the {@linkplain MessageClass MessageClass} of {@code |M|}.
+     * @param <R> type of an error code to use for error reporting; must be a Protobuf enum value
+     */
+    @Internal
+    public abstract static class ExceptionBuilder<E extends Exception,
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: It looks more like a factory.

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java [hunk 6]
```diff
@@ -128,4 +137,89 @@ public static String toText(String format, Iterable<ConstraintViolation> violati
...
         return resultBuilder.toString();
     }
+
+    /**
+     * A helper class for building exceptions used to report invalid {@code Message}s,
+     * which have fields that violate validation constraint(s).
+     *
+     * @param <E> type of {@code Exception} to build
+     * @param <M> type of the {@code Message}
+     * @param <C> type of the {@linkplain MessageClass MessageClass} of {@code |M|}.
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: > {@linkplain MessageClass MessageClass}

Why cannot we just `{@link MessageClass}`?

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java [hunk 7]
```diff
@@ -128,4 +137,89 @@ public static String toText(String format, Iterable<ConstraintViolation> violati
...
+        protected abstract C getMessageClass();
+
+        /**
+         * Obtains an error code to use for error reporting.
+         */
+        protected abstract R getErrorCode();
+
+        protected abstract String getErrorText();
+
+
```
Comment (alexander-yevsyukov) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: Please remove empty line.

File: server/src/main/java/org/spine3/server/event/InvalidEventException.java [hunk 8]
```diff
@@ -56,31 +58,54 @@ private InvalidEventException(String messageText, Message eventMsg, Error error)
...
+            return MSG_VALIDATION_ERROR;
+        }
+
+        @Override
+        protected Map<String, Value> getMessageTypeAttribute(Message message) {
+            return eventTypeAttribute(message);
+        }
+
+        @Override
+        protected InvalidEventException createException(String exceptionMsg, Message message,
```
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/event/InvalidEventException.java]: Please fix wrapping of arguments.

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java (hunk 6 above)
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: My bad. Fixed.

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java (hunk 5 above)
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: Renamed. Its method has been renamed from `buildException` to `newException` as well.

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java (hunk 7 above)
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: Done.

File: client/src/main/java/org/spine3/validate/ConstraintViolations.java (hunk 1 above)
Comment (armiol) [inline client/src/main/java/org/spine3/validate/ConstraintViolations.java]: There are 16 more usages of `String.format()` (20 in total)  in our code. 
And 22 statically imported use-cases. 

Let's discuss and fix as a separate PR.

File: server/src/main/java/org/spine3/server/command/InvalidCommandException.java (hunk 2 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/command/InvalidCommandException.java]: Of course. Fixed.

File: server/src/main/java/org/spine3/server/command/InvalidCommandException.java (hunk 3 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/command/InvalidCommandException.java]: Rearranged.

File: server/src/main/java/org/spine3/server/event/InvalidEventException.java (hunk 4 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/InvalidEventException.java]: Same as above. Fixed.

File: server/src/main/java/org/spine3/server/event/InvalidEventException.java (hunk 8 above)
Comment (armiol) [inline server/src/main/java/org/spine3/server/event/InvalidEventException.java]: Fixed.

**Comment by armiol** [comment]
@alexander-yevsyukov the comments addressed. PTAL.

**Pull request merged**